storage/src/context.rs

```diff
@@ -1,218 +1,73 @@
-
-    /// Checks context and resolves keys to be delete a place them to diff, and also deletes keys from diff
-    fn delete_to_diff(&self, context_hash: &Option<ContextHash>, key_prefix_to_delete: &Vec<String>, context_diff: &mut ContextDiff) -> Result<(), ContextError>;
-
-    /// Checks context and resolves keys to be delete a place them to diff, and also deletes keys from diff
-    fn remove_recursively_to_diff(&self, context_hash: &Option<ContextHash>, key_prefix_to_remove: &Vec<String>, context_diff: &mut ContextDiff) -> Result<(), ContextError>;
-
-    /// Checks context and copies subtree under 'from_key' to new subtree under 'to_key'
-    fn copy_to_diff(&self, context_hash: &Option<ContextHash>, from_key: &Vec<String>, to_key: &Vec<String>, context_diff: &mut ContextDiff) -> Result<(), ContextError>;
-
-    fn get_key(&self, context_index: &ContextIndex, key: &Vec<String>) -> Result<Option<Bucket<Vec<u8>>>, ContextError>;
-}
-
-fn to_key(key: &Vec<String>) -> String {
-    key.join("/")
-}
-
-fn key_terminated_starts_with(key: &String, prefix: &Vec<String>) -> bool {
-    let mut terminated_key = to_key(prefix);
-    // terminate the key with a '/'
-    terminated_key.push('/');
-    key.starts_with(&terminated_key)
-}
-
-fn replace_key(key: &String, matched: &Vec<String>, replacer: &Vec<String>) -> String {
-    key.replace(&to_key(matched), &to_key(replacer))
-}
-
-/// Struct points to context commmit hash or index, which is checkouted
-pub struct ContextIndex {
-    level: Option<usize>,
-    pub context_hash: Option<ContextHash>,
-}
-
-impl ContextIndex {
-    pub fn new(level: Option<usize>, context_hash: Option<ContextHash>) -> Self {
-        ContextIndex { level, context_hash }
-    }
+    // set key-value
+    fn set(&mut self, context_hash: &Option<ContextHash>, key: &[String], value: &[u8]) -> Result<(), ContextError>;
+    // checkout context for hash
+    fn checkout(&self, context_hash: &ContextHash) -> Result<(), ContextError>;
+    // commit current context diff to storage
+    // if parent_context_hash is empty, it means that it's a commit_genesis and we don't assign context_hash to header
+    fn commit(&mut self, block_hash: &BlockHash, parent_context_hash: &Option<ContextHash>,
+              new_context_hash: &ContextHash, author: String, message: String,
+              date: i64) -> Result<(), ContextError>;
+    fn delete_to_diff(&self, context_hash: &Option<ContextHash>, key_prefix_to_delete: &[String]) -> Result<(), ContextError>;
+    fn remove_recursively_to_diff(&self, context_hash: &Option<ContextHash>, key_prefix_to_remove: &[String]) -> Result<(), ContextError>;
+    // copies subtree under 'from_key' to new subtree under 'to_key'
+    fn copy_to_diff(&self, context_hash: &Option<ContextHash>, from_key: &[String], to_key: &[String]) -> Result<(), ContextError>;
+    // get value for key
+    fn get_key(&self, key: &Vec<String>) -> Result<Vec<u8>, ContextError>;
```

Comment:
@michal-stanek 
what is the best way?
`key: &Vec<String>` vs `key: &[String]` vs `prefix: &ContextKey`

I think this is a trait/api and we should align this somehow

Reply:
I think &ContextKey would be best for both key and prefix. I will change it.